src/rosdep2/main.py

```diff
@@ -391,6 +392,18 @@ def _rosdep_main(args):
         check_for_sources_list_init(options.sources_cache_dir)
     elif command not in ['fix-permissions']:
         setup_proxy_opener()
+
+    if 'ROS_PYTHON_VERSION' not in os.environ and options.ros_distro:
```

Comment:
If the ROS distro has information about the Python version I would expect it to be used and override the one set in the environment. Otherwise releasing a package for Noetic when Melodic is sourced will result in incorrect Debian control files.

Reply:
I think this PR only affects the CLI tool. Bloom still needs to pick the right Python version when generating Debian control files.

Reply:
So what happens when I have Melodic sourced but created a Noetic workspace and trying to install its dependencies using `rosdep` and explicitly set `--rosdistro noetic`?

Reply:
> So what happens when I have Melodic sourced but created a Noetic workspace and trying to install its dependencies using rosdep and explicitly set --rosdistro noetic?

All the bad things one would expect. If the melodic workspace has the `ros_environment` package, the environment variable `ROS_PYTHON_VERSION` would be set to `2`. Conditional dependencies in the Noetic workspace would be evaluated using `ROS_PYTHON_VERSION=2`, which is opposite of the intent of the package authors in the Noetic workspace. With #709 pip dependencies would be installed with Python 2 as well. This could be a problem for users because it's easy to accidentally have a `melodic` workspace sourced (like, if it's been added to `~/.bashrc`).

 Is there something we can do to address this case without breaking platforms that jumped to Python 3 already? Keeping with the same scenario of a user having a Melodic workspace sourced, say they manage to install all the right dependencies of their new Noetic workspace, and now they're ready to run `catkin_make`. What does `catkin_make` do to prevent packages in the `Noetic` workspace from using package in the `Melodic` workspace. Maybe we can do the same thing here?

Reply:
> What does `catkin_make` do to prevent packages in the `Noetic` workspace from using package in the `Melodic` workspace?

Nothing. For `catkin` the already sourced workspace is just an underlay which it will happily fall back to.

Only the `ros_environment` package does some sanity checking and warning the user if they mix ROS distros: see https://github.com/ros/ros_environment/blob/7ede5949279af45582aefd124d6d69a9125d985e/env-hooks/1.ros_distro.sh.in#L4

On a CMake level `catkin` will simply use the `ROS_PYTHON_VERSION` env var (see https://github.com/ros/catkin/blob/77822d31cd84c052e776f58c4d5f5630e0760725/cmake/python.cmake#L3) - in this case Python 2 for Noetic :boom:

Reply:
I think we can do both. Thoughts on dd4bf62? If `--rosdistro` and `ROS_DISTRO` differ, then rosdep prefers `--rosdistro` and the index over the environment. If `--rosdistro` is unspecified or the same as `ROS_DISTRO`, then rosdep prefers the environment over the index.

In the case of a user with a melodic workspace sourced trying to get dependencies for a noetic workspace, rosdep will use Python 3 since that's in the index for noetic.

```
(rosdep-env3) sloretz@nxt /t/urdf_parser_py> env ROS_DISTRO=melodic ROS_PYTHON_VERSION=2 rosdep check --from-paths . --skip-keys="rospy catkin" --rosdistro noetic
WARNING: given --rosdistro noetic but ROS_DISTRO is "melodic". Ignoring environment.
System dependencies have not been satisfied:
apt	python3-lxml
apt	python3-mock
```

In the case of a user on an unsupported platform using melodic and Python 3 to get dependencies for a melodic workspace, rosdep will use Python 3 since that's the version in the environment variable.


```
(rosdep-env3) sloretz@nxt /t/urdf_parser_py> env ROS_DISTRO=melodic ROS_PYTHON_VERSION=3 rosdep check --from-paths . --skip-keys="rospy catkin"
System dependencies have not been satisfied:
apt	python3-lxml
apt	python3-mock
(rosdep-env3) sloretz@nxt /t/urdf_parser_py> env ROS_DISTRO=melodic ROS_PYTHON_VERSION=3 rosdep check --from-paths . --skip-keys="rospy catkin" --rosdistro melodic
System dependencies have not been satisfied:
apt	python3-lxml
apt	python3-mock
```

And checking normal cases, melodic sourced fetching dependencies for a melodic workspace always uses python 2
```
(rosdep-env3) sloretz@nxt /t/urdf_parser_py> env ROS_DISTRO=melodic ROS_PYTHON_VERSION=2 rosdep check --from-paths . --skip-keys="rospy catkin"
All system dependencies have been satisfied
(rosdep-env3) sloretz@nxt /t/urdf_parser_py> env ROS_DISTRO=melodic ROS_PYTHON_VERSION=2 rosdep check --from-paths . --skip-keys="rospy catkin" --rosdistro melodic
All system dependencies have been satisfied
```

noetic sourced fetching dependencies for a noetic workspace always uses python 3
```
(rosdep-env3) sloretz@nxt /t/urdf_parser_py> env ROS_DISTRO=noetic ROS_PYTHON_VERSION=3 rosdep check --from-paths . --skip-keys="rospy catkin"
System dependencies have not been satisfied:
apt	python3-lxml
apt	python3-mock
(rosdep-env3) sloretz@nxt /t/urdf_parser_py> env ROS_DISTRO=noetic ROS_PYTHON_VERSION=3 rosdep check --from-paths . --skip-keys="rospy catkin" --rosdistro noetic
System dependencies have not been satisfied:
apt	python3-lxml
apt	python3-mock
```

No workspace sourced always uses `ROS_PYTHON_VERSION`

```
(rosdep-env3) sloretz@nxt /t/urdf_parser_py> env ROS_PYTHON_VERSION=3 rosdep check --from-paths . --skip-keys="rospy catkin" --rosdistro noetic
System dependencies have not been satisfied:
apt	python3-lxml
apt	python3-mock
(rosdep-env3) sloretz@nxt /t/urdf_parser_py> env ROS_PYTHON_VERSION=2 rosdep check --from-paths . --skip-keys="rospy catkin" --rosdistro noetic
All system dependencies have been satisfied
```

Reply:
Sounds like a very good heuristic to me :+1: